src/simulation_strategies/generalized_cliffords_b_simulation_strategy.py

```diff
@@ -0,0 +1,107 @@
+__author__ = "Tomasz Rybotycki"
+
+"""
+    This script contains the implementation of the Generalized Cliffords version B
+    algorithm. It differs from the original version  in two places:
+        -   In the k-th step, instead of computing k x k permanents, we compute a set of 
+            k-1 x k-1 permanents. From that, we calculate the required k x k permanents
+            in O(k) time each.
+        -   We permute the columns of the matrix at the beginning. We can then sample
+            from an easier distribution and obtain the same results.
+"""
+
+from .generalized_cliffords_simulation_strategy import GeneralizedCliffordsSimulationStrategy, BSPermanentCalculatorInterface
+from numpy import array, ndarray, int64
+from typing import List
+from numpy.random import choice
+from copy import deepcopy
+
+class GeneralizedCliffordsBSimulationStrategy(GeneralizedCliffordsSimulationStrategy):
+
+    def __init__(self, bs_permanent_calculator: BSPermanentCalculatorInterface) -> None:
+        self._current_input = []
+        self._working_input_state = None
+        super().__init__(bs_permanent_calculator)
+
+    def simulate(self, input_state: ndarray, samples_number: int = 1) -> List[ndarray]:
+        """
+            Returns sample from linear optics experiments given output state.
+
+            :param input_state: Input state in particle basis.
+            :param samples_number: Number of samples to simulate.
+            :return: A resultant state after traversing through interferometer.
+        """
+        self.input_state = input_state
+        self.number_of_input_photons = sum(input_state)
+        self.pmfs = dict()
+
+        samples = []
+
+        while len(samples) < samples_number:
+            self._current_input = [0 for _ in self.input_state]
```

Comment:
Does it have to be `list`? How often is this called? Maybe using `numpy` `zeros` would make it perform better? Do we care?